### PR TITLE
Fixed #26936 -- Fixed stale ContentType deletion if an app doesn't have any models.

### PR DESCRIPTION
**Pull request by ChaosOrd**
https://code.djangoproject.com/ticket/26936

### CODE REVIEW

**Comment by timgraham** [comment]
Using `TestCase.available_apps` may be useful here. I see that when running the full test suite, the new test deletes and recreates 1459 content types. This is likely not too fast and is probably the source of the test isolation issues.

**Comment by ChaosOrd** [comment]
Rewrote the test to avoid the isolation issues

File: tests/contenttypes_tests/test_management.py [hunk 1]
```diff
@@ -1,16 +1,23 @@
...
 from .models import ModelWithNullFKToSite, Post
 
 
 class UpdateContentTypesTests(TestCase):
+    available_apps = [
+        'contenttypes_tests',
+        'contenttypes_tests.app_with_no_models',
+        'django.contrib.contenttypes',
+    ]
+    apps.set_installed_apps(settings.INSTALLED_APPS + ['contenttypes_tests.app_with_no_models'])
```
Comment (timgraham) [inline tests/contenttypes_tests/test_management.py]: What's this line?  Looks similar to what `TransactionTestCase._pre_setup()` does -- I don't think both should be needed and writing code that executes in the class body is a suspicious..

File: tests/contenttypes_tests/test_management.py [hunk 2]
```diff
@@ -63,3 +70,12 @@ def test_unavailable_content_type_model(self):
...
             contenttypes_management.create_contenttypes(self.app_config, interactive=False, verbosity=0, apps=apps)
         self.assertEqual(ContentType.objects.count(), self.before_count + 1)
+
+    def test_all_models_removed(self):
+        ContentType.objects.create(app_label='app_with_no_models', model='Fake')
+        with mock.patch('builtins.input', return_value='yes'):
+            with captured_stdout() as stdout:
+                call_command('remove_stale_contenttypes', verbosity=2)
+
+            self.assertIn("Deleting stale content type", stdout.getvalue())
```
Comment (timgraham) [inline tests/contenttypes_tests/test_management.py]: Use single quotes unless the string contains a single quote.  Check for the correct content type that's deleted?

File: tests/contenttypes_tests/test_management.py [hunk 3]
```diff
@@ -63,3 +70,12 @@ def test_unavailable_content_type_model(self):
...
+            self.assertEquals(ContentType.objects.count(), self.before_count)
```
Comment (timgraham) [inline tests/contenttypes_tests/test_management.py]: assertEqual -- the version with "s" is a deprecated alias.

File: tests/contenttypes_tests/test_management.py [hunk 4]
```diff
@@ -1,16 +1,23 @@
 from unittest import mock
 
 from django.apps.registry import Apps, apps
+from django.conf import settings
 from django.contrib.contenttypes import management as contenttypes_management
 from django.contrib.contenttypes.models import ContentType
 from django.core.management import call_command
 from django.test import TestCase
 from django.test.utils import captured_stdout
-
```
Comment (timgraham) [inline tests/contenttypes_tests/test_management.py]: revert

File: tests/contenttypes_tests/test_management.py [hunk 5]
```diff
@@ -11,6 +12,17 @@
...
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.available_apps = [
+            'contenttypes_tests',
+            'contenttypes_tests.app_with_no_models',
+            'django.contrib.contenttypes',
+        ]
+        apps.set_installed_apps(settings.INSTALLED_APPS + ['contenttypes_tests.app_with_no_models'])
```
Comment (ChaosOrd) [inline tests/contenttypes_tests/test_management.py]: Could not really avoid putting those lines, though I moved them to setUpClass from the class body.
The command that deletes the content types iterates over available apps. that's why I need the first line. "available_apps" needs to be a subset of "installed_apps", therefore a second line.

Is there a better way to do it?

File: tests/contenttypes_tests/test_management.py (hunk 1 above)
Comment (ChaosOrd) [inline tests/contenttypes_tests/test_management.py]: Could not really avoid putting those lines. The command that deletes the content types iterates over available apps. that's why I need the first line. "available_apps" needs to be a subset of "installed_apps", therefore a second line.

Is there a better way doing it?

Comment (ChaosOrd) [inline tests/contenttypes_tests/test_management.py]: Is mocking the available_apps is acceptable in this case?

**Comment by ChaosOrd** [comment]
Addressed all the points from the review except one. (it is not clear what is the alternative, please see my comment in the code)

**Comment by timgraham** [comment]
I found that `apps.set_installed_apps()` can be avoided by decorating `UpdateContentTypesTests` with  `@modify_settings(INSTALLED_APPS={'append': ['no_models']})` (I switched 'contenttypes_tests.app_with_no_models' to 'no_models'. Now I'm trying to remember why `available_apps` is needed -- do you remember? Was the test rewritten since my first comment such that it's not needed anymore?

**Pull request merged**